### PR TITLE
feat(APIM-471): endpoint GET /customers - get information from Salesforce through Informatica REST API

### PR DESCRIPTION
**Pull request by avaitonis**
### Introduction
New endpoint to get Customer information from Salesforce trough Informatica REST API.
Informatica provides main functionality, this is just wrapper to have consistent authentication, monitoring and easy access.
Customer search is possible using one of three query parameters: `companyreg`, `partyUrn`, `name`.

### Resolution
- Copied http module from Estore API. 
- Module Informatica is using http module and is based on Custodian module from Estore API. Maybe not latest version?
- Added logging and testing pieces from our other APIs. 
  - Not everything is tested.
- Renamed query parameter `includeLegacyData` to `fallbackToLegacyData`.
- We use basic auth to communicate with Informatica.

### Todo
- Add log redacting
- Test if http module was copied correct and all logging is working.
- Informatica Rest API has SSL certificate issues when running in local network.
- New environment variables needs to be added to Github.

### CODE REVIEW

File: src/modules/customers/customers.controller.test.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+      {
+        query: { name: CUSTOMERS.EXAMPLES.NAME, partyUrn: CUSTOMERS.EXAMPLES.PARTYURN },
+      },
+      {
+        query: { companyreg: CUSTOMERS.EXAMPLES.COMPANYREG, partyUrn: CUSTOMERS.EXAMPLES.PARTYURN },
+      },
+      {
+        query: { name: CUSTOMERS.EXAMPLES.NAME, companyreg: CUSTOMERS.EXAMPLES.COMPANYREG, partyUrn: CUSTOMERS.EXAMPLES.PARTYURN },
+      },
+    ])('throws BadRequestException if search parameter (name, companyreg, partyUrn) number is not 1', ({ query }) => {
```
Comment (lmarrai-sw) [inline src/modules/customers/customers.controller.test.ts:95]: `'throws BadRequestException if there is more than 1 search parameter (name, companyreg, partyUrn)'` might be a bit clearer

File: src/modules/customers/dto/get-customers-query.dto.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+export class GetCustomersQueryDto {
+  @IsOptional()
+  @ApiProperty({
+    required: false,
+    example: CUSTOMERS.EXAMPLES.COMPANYREG,
+    description: 'The company registration number (in the country where the customer is registered) of the customer to search for.',
+  })
+  @MinLength(8)
+  @MaxLength(10)
+  public companyreg?: string;
```
Comment (lmarrai-sw) [inline src/modules/customers/dto/get-customers-query.dto.ts:15]: In the ticket, we spelt this as `companyReg` instead - did we want to change that? (I prefer `companyReg` myself as it matches the response body)

**Review by lmarrai-sw — CHANGES_REQUESTED** [review]
Great work here @avaitonis, thank you

File: .env.sample [hunk 3]
```diff
@@ -35,5 +37,12 @@ REDIS_TTL=100
 REDIS_HOST=localhost
 REDIS_PORT=6379
 
+# INFORMATICA Rest API
+APIM_INFORMATICA_URL=
+APIM_INFORMATICA_USERNAME=
+APIM_INFORMATICA_PASSWORD=
+APIM_INFORMATICA_MAX_REDIRECTS=
+APIM_INFORMATICA_TIMEOUT=
```
[the remark this answers is not in the document]
Comment (avaitonis) [reply .env.sample:45]: Created ticket https://ukef-dtfs.atlassian.net/browse/APIM-526

File: package.json [hunk 4]
```diff
@@ -51,6 +51,7 @@
   "devDependencies": {
     "@commitlint/cli": "^17.6.5",
     "@commitlint/config-conventional": "^17.6.5",
+    "@nestjs/axios": "^3.0.0",
```
[the remark this answers is not in the document]
Comment (avaitonis) [reply package.json]: Thanks, moved to normal dependencies.

File: package.json [hunk 5]
```diff
@@ -63,6 +64,7 @@
     "@types/supertest": "^2.0.12",
     "@typescript-eslint/eslint-plugin": "^5.61.0",
     "@typescript-eslint/parser": "^5.61.0",
+    "axios": "^1.4.0",
```
[the remark this answers is not in the document]
Comment (avaitonis) [reply package.json]: Thanks, moved to normal dependencies.

File: test/setup/override-environment-variables.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import { getEnvironmentVariablesForProcessEnv } from '@ukef-test/support/environment-variables';
+
+let originalProcessEnv: NodeJS.ProcessEnv;
+
+beforeAll(() => {
+  originalProcessEnv = process.env;
+  // Override some environment variables.
+  process.env = { ...process.env, ...getEnvironmentVariablesForProcessEnv() };
```
[the remark this answers is not in the document]
Comment (avaitonis) [reply test/setup/override-environment-variables.ts:8]: Most endpoints are pointing to one of 4 DBs, I didn't move them to test ENV variables.
Other api-tests can point to Azure dev or docker test DB, so for now it's not decided.

File: test/customers/get-customers.api-test.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
+// import { PROPERTIES } from '@ukef/constants';
+// import { withAcbsAuthenticationApiTests } from '@ukef-test/common-tests/acbs-authentication-api-tests';
```
[the remark this answers is not in the document]
Comment (avaitonis) [reply test/customers/get-customers.api-test.ts]: Deleted

File: src/modules/customers/customers.controller.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import { GetCustomersResponse, GetCustomersResponseItem } from './dto/get-customers-response.dto';
+
+@ApiTags('customers')
+@Controller('customers')
+export class CustomersController {
+  constructor(private readonly customersService: CustomersService) {}
+
+  @Get()
+  @ApiOperation({
+    summary: 'Get customers from salesforce',
```
[the remark this answers is not in the document]
Comment (avaitonis) [reply src/modules/customers/customers.controller.ts]: Done

File: src/modules/customers/customers.controller.test.ts (hunk 1 above)
Comment (avaitonis) [reply src/modules/customers/customers.controller.test.ts]: Changed

File: src/modules/customers/dto/get-customers-query.dto.ts (hunk 2 above)
Comment (avaitonis) [reply src/modules/customers/dto/get-customers-query.dto.ts]: Agree, changed companyReg, but in response it is companyRegNo

**Review by lmarrai-sw — APPROVED** [review]
LGTM, thanks @avaitonis

**Pull request merged**